[release-8.3] Fix hang in file watcher

monojenkins opened the pull request:
Enabling events has a sync wait on a flush of the IO event queue, and
that may take many seconds to complete if there is a lot of IO. To
avoid locking on this call, we now enable the events on a background
thread.

Fixes VSTS #952405 - Visual Studio Mac 2019 Freezes on Switch Branch
Fixes VSTS #980470 - Deadlock in FileWatcherService.WatchDirectories

Backport of #8688.

/cc @slluis

slluis committed:
Fix hang in file watcher

Enabling events has a sync wait on a flush of the IO event queue, and
that may take many seconds to complete if there is a lot of IO. To
avoid locking on this call, we now enable the events on a background
thread.

Fixes VSTS #952405 - Visual Studio Mac 2019 Freezes on Switch Branch
Fixes VSTS #980470 - Deadlock in FileWatcherService.WatchDirectories

Therzok reviewed (changes requested):
See other PR for comments.

slluis committed:
Fix FSW update logic

When updating watchers, if a watcher already exists ensure it is fully
initialized before leaving the Update method.
